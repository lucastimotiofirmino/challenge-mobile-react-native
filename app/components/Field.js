import React, { memo } from 'react';

import { StyleSheet, TouchableOpacity, Text, View } from 'react-native';

import Icon from 'react-native-vector-icons/AntDesign';

const styles = StyleSheet.create({
  field: {
    borderRadius: 25,
    paddingHorizontal: 20,
    paddingVertical: 8,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    marginRight: 5,
  },
  fieldIcon: {
    color: '#fff',
    fontSize: 18,
    marginRight: 10,
  },
  fieldText: {
    fontWeight: 'bold',
    color: '#fff',
    fontSize: 13,
    textAlign: 'center',
  },
});

const Field = ({ icon, text, onPress, backgroundColor = '#F67571' }) => {
  const content = (
    <View style={[styles.field, { backgroundColor }]} testID="fieldbutton">
      <Icon name={icon} style={styles.fieldIcon} />
      <Text style={styles.fieldText} numberOfLines={1}>
        {text}
      </Text>
    </View>
  );

  if (onPress !== undefined) {
    return (
      <TouchableOpacity onPress={() => onPress()}>{content}</TouchableOpacity>
    );
  }

  return content;
};

export default memo(Field);
