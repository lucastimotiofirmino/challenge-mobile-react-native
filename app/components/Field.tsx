import React, { memo } from 'react';

import {
  StyleSheet,
  TouchableOpacity,
  Text,
  View,
  ViewStyle,
  TextStyle,
} from 'react-native';

import Icon from 'react-native-vector-icons/AntDesign';

type Style = {
  field: ViewStyle;
  fieldIcon: ViewStyle;
  fieldText: TextStyle;
};

const styles = StyleSheet.create<Style>({
  field: {
    borderRadius: 25,
    paddingHorizontal: 20,
    paddingVertical: 8,
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'row',
    marginRight: 5,
  },
  fieldIcon: {
    color: '#fff',
    fontSize: 18,
    marginRight: 10,
  },
  fieldText: {
    fontWeight: 'bold',
    color: '#fff',
    fontSize: 13,
    textAlign: 'center',
  },
});

type FieldProps = {
  icon: string;
  text: string;
  backgroundColor?: string;
  onPress?: Function | null;
};

const Field: React.FC<FieldProps> = ({
  icon,
  text,
  onPress = null,
  backgroundColor = '#F67571',
}) => {
  const content = (
    <View style={[styles.field, { backgroundColor }]} testID="fieldbutton">
      <Icon name={icon} style={styles.fieldIcon} />
      <Text style={styles.fieldText} numberOfLines={1}>
        {text}
      </Text>
    </View>
  );

  if (onPress !== null) {
    return (
      <TouchableOpacity onPress={() => onPress()}>{content}</TouchableOpacity>
    );
  }

  return content;
};

export default memo(Field);
