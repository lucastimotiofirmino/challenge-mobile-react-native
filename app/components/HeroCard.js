import React, { memo } from 'react';

import {
  StyleSheet,
  Image,
  TouchableNativeFeedback,
  Text,
  View,
  TouchableOpacity,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { useDispatch } from 'react-redux';
import { addHeroFavorite, removeHeroFavorite } from '../actions/favorite';
import Icon from 'react-native-vector-icons/AntDesign';

export const CARD_HEIGHT = 300; // utilized by the FlatList for optimization

const styles = StyleSheet.create({
  card: {
    borderRadius: 10,
    marginHorizontal: 5,
    backgroundColor: '#ededed',
    marginBottom: 15,
    shadowColor: '#000',
    shadowOffset: {
      width: 1,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 0.5,
    elevation: 2,
  },
  name: {
    maxWidth: 140,
    marginVertical: 20,
    fontSize: 20,
    fontWeight: 'bold',
    textAlign: 'center',
    alignSelf: 'center',
  },
  photo: {
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    height: 200,
    width: 180,
  },
  modalOpenIcon: {
    backgroundColor: 'rgba(0, 0, 0, 0.35)',
    borderRadius: 20,
    top: 5,
    padding: 5,
    right: 5,
    position: 'absolute',
    zIndex: 99,
    color: '#fff',
    fontSize: 18,
  },
  actionButton: {
    backgroundColor: '#FAA367',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  actionText: {
    color: '#333',
    marginLeft: 8,
  },
});

const HeroCard = ({ ...hero }) => {
  const { name, avatar, favorite } = hero;
  const navigation = useNavigation();
  const dispatch = useDispatch();

  const handleFavorite = () => {
    if (!favorite) {
      dispatch(addHeroFavorite({ hero }));
      return;
    }

    dispatch(removeHeroFavorite({ hero }));
  };

  const card = (
    <View style={styles.card}>
      <Icon name="arrowsalt" style={styles.modalOpenIcon} />
      <Image source={{ uri: avatar }} style={styles.photo} />
      <TouchableOpacity
        style={styles.actionButton}
        onPress={() => handleFavorite()}>
        <Icon
          name={favorite ? 'star' : 'staro'}
          color={favorite ? '#333' : '#333'}
          size={18}
        />
        <Text style={styles.actionText}>
          {favorite ? 'Desfavoritar' : 'Favoritar'}
        </Text>
      </TouchableOpacity>
      <Text style={styles.name}>{name}</Text>
    </View>
  );

  const openHeroModal = () => {
    navigation.navigate('HeroModal', hero);
  };

  return (
    <TouchableNativeFeedback onPress={() => openHeroModal()} testID="cardHero">
      {card}
    </TouchableNativeFeedback>
  );
};

export default memo(HeroCard);
