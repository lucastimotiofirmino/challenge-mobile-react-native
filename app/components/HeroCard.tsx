import React, { memo } from 'react';

import {
  StyleSheet,
  Image,
  Text,
  View,
  TouchableOpacity,
  TouchableNativeFeedback,
  Dimensions,
  ViewStyle,
  TextStyle,
  ImageStyle,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import { useDispatch } from 'react-redux';
import { addHeroFavorite, removeHeroFavorite } from '../actions/favorite';
import Icon from 'react-native-vector-icons/AntDesign';
import { IHero } from '../common/interfaces';

const { width } = Dimensions.get('screen');

export const CARD_HEIGHT: number = 300; // utilized by the FlatList for optimization

type Style = {
  card: ViewStyle;
  name: TextStyle;
  photo: ImageStyle;
  modalOpenContainer: ViewStyle;
  modalOpenIcon: TextStyle;
  actionButton: ViewStyle;
  actionText: TextStyle;
};

const styles = StyleSheet.create<Style>({
  card: {
    borderRadius: 10,
    marginHorizontal: 5,
    backgroundColor: '#ededed',
    marginBottom: 15,
    shadowColor: '#000',
    shadowOffset: {
      width: 1,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 0.5,
    elevation: 2,
  },
  name: {
    maxWidth: 140,
    marginVertical: 20,
    fontSize: 20,
    fontWeight: 'bold',
    textAlign: 'center',
    alignSelf: 'center',
  },
  photo: {
    borderTopLeftRadius: 10,
    borderTopRightRadius: 10,
    width: width / 2 - 20,
    height: 180,
  },
  modalOpenContainer: {
    position: 'absolute',
    top: 10,
    right: 10,
    zIndex: 9,
    backgroundColor: 'rgba(0, 0, 0, 0.35)',
    borderRadius: 20,
    padding: 8,
  },
  modalOpenIcon: {
    color: '#fff',
    fontSize: 18,
  },
  actionButton: {
    backgroundColor: '#FAA367',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
  },
  actionText: {
    color: '#333',
    marginLeft: 8,
  },
});

const HeroCard: React.FC<IHero> = ({...hero}) => {
  const { name, avatar, favorite } = hero;
  const navigation = useNavigation();
  const dispatch = useDispatch();

  const handleFavorite = () => {
    if (!favorite) {
      dispatch(addHeroFavorite(hero));
      return;
    }

    dispatch(removeHeroFavorite(hero));
  };

  const card = (
    <View style={styles.card}>
      <View style={styles.modalOpenContainer}>
        <Icon name="arrowsalt" style={styles.modalOpenIcon} />
      </View>
      <Image source={{ uri: avatar }} style={styles.photo} />
      <TouchableOpacity
        style={styles.actionButton}
        onPress={() => handleFavorite()}>
        <Icon
          name={favorite ? 'star' : 'staro'}
          color={favorite ? '#333' : '#333'}
          size={18}
        />
        <Text style={styles.actionText}>
          {favorite ? 'Desfavoritar' : 'Favoritar'}
        </Text>
      </TouchableOpacity>
      <Text style={styles.name}>{name}</Text>
    </View>
  );

  const openHeroModal = () => {
    navigation.navigate('HeroModal', hero);
  };

  return (
    <TouchableNativeFeedback onPress={() => openHeroModal()} testID="cardHero">
      {card}
    </TouchableNativeFeedback>
  );
};

export default memo(HeroCard);
