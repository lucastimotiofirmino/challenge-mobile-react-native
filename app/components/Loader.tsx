import React, { useEffect } from 'react';

import { StyleSheet, Animated, Easing, ImageStyle } from 'react-native';
import icon from '../../assets/captain-icon.png';

const LOADER_SIZE: number = 42;

type Style = {
  loader: ImageStyle;
};

const styles = StyleSheet.create<Style>({
  loader: {
    width: LOADER_SIZE,
    height: LOADER_SIZE,
    alignSelf: 'center',
    marginVertical: 10,
  },
});

type LoaderProps = {
  style?: ImageStyle;
};

const Loader: React.FC<LoaderProps> = ({ style }) => {
  const spinAnimationValue = new Animated.Value(0);

  useEffect(() => {
    Animated.loop(
      Animated.timing(spinAnimationValue, {
        toValue: 1,
        duration: 1000,
        easing: Easing.bezier(0.645, 0.045, 0.355, 1.0),
        useNativeDriver: true,
        isInteraction: false,
      }),
    ).start();
  }, []);

  const spinAnimationRotate = spinAnimationValue.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg'],
  });

  return (
    <Animated.Image
      source={icon}
      style={{
        transform: [{ rotate: spinAnimationRotate }],
        ...styles.loader,
        ...style,
      }}
    />
  );
};

export default Loader;
