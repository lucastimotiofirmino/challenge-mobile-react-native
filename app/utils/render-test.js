import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from '../common/store';

export function renderWithRedux(ui, { store, ...otherOpts }) {
  return render(<Provider store={store}>{ui}</Provider>, otherOpts);
}

export function makeTestStore(opts = {}) {
  const { store } = createStore(opts);
  return store;
}
