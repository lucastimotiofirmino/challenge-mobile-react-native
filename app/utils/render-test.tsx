import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import createStore from '../common/store';
import { Store } from 'redux';

export function renderWithRedux(
  ui: React.ReactNode,
  { store, ...otherOpts }: { store: Store },
) {
  return render(<Provider store={store}>{ui}</Provider>, otherOpts);
}

export function makeTestStore() {
  const { store } = createStore();
  return store;
}
