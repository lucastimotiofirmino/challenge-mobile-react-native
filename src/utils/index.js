const prepareAction = str => {
    return {
        ACTION: str,
        SUCCESS: `SUCCESS_${str}`,
        FAILED: `FAILED_${str}`,
    }
}

const handleRes = res => {
    if (!res.ok) {
        console.log(`Error ${res.status}: ${res.statusText}`)
        throw new Error(res.statusText)
    }
    return res.json()
}

export {prepareAction, handleRes}
