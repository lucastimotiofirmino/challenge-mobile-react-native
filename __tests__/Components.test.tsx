import 'react-native';
import React from 'react';
import { render, cleanup } from '@testing-library/react-native';
import Loader from '../app/components/Loader';
import Field from '../app/components/Field';
import HeroCard from '../app/components/HeroCard';
import { renderWithRedux, makeTestStore } from '../app/utils/render-test';

describe('suite simple components renders', () => {
  afterEach(cleanup);

  test('renders correctly the custom loader component', () => {
    const loader = render(<Loader />).toJSON();

    expect(loader?.type).toBe('Image');
    expect(loader).toMatchSnapshot();
  });

  test('renders correctly the hero card component', () => {
    const hero = {
      id: "1",
      name: 'Wolverine',
      avatar: 'test',
      numEvents: 0,
    };
    const { getByText, toJSON } = renderWithRedux(<HeroCard {...hero} />, {store: makeTestStore()});

    expect(getByText(/Wolverine/i)).toBeTruthy();

    expect(toJSON()).toMatchSnapshot();
  });

  test('renders correctly the field component', () => {
    const field = {
      icon: 'down',
      text: 'TestContentField',
    };
    const { getByText, toJSON } = render(<Field {...field} />);

    expect(getByText(/TestContentField/i)).toBeTruthy();

    expect(toJSON()).toMatchSnapshot();
  });
});
