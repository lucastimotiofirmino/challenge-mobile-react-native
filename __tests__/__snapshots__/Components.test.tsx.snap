// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`suite simple components renders renders correctly the custom loader component 1`] = `
<Image
  source={
    Object {
      "process": [Function],
    }
  }
  style={
    Object {
      "alignSelf": "center",
      "height": 42,
      "marginVertical": 10,
      "transform": Array [
        Object {
          "rotate": "0deg",
        },
      ],
      "width": 42,
    }
  }
/>
`;

exports[`suite simple components renders renders correctly the field component 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderRadius": 25,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginRight": 5,
        "paddingHorizontal": 20,
        "paddingVertical": 8,
      },
      Object {
        "backgroundColor": "#F67571",
      },
    ]
  }
  testID="fieldbutton"
>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Object {
          "color": "#fff",
          "fontSize": 18,
          "marginRight": 10,
        },
        Object {
          "fontFamily": "anticon",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "#fff",
        "fontSize": 13,
        "fontWeight": "bold",
        "textAlign": "center",
      }
    }
  >
    TestContentField
  </Text>
</View>
`;

exports[`suite simple components renders renders correctly the hero card component 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#ededed",
      "borderRadius": 10,
      "elevation": 2,
      "marginBottom": 15,
      "marginHorizontal": 5,
      "shadowColor": "#000",
      "shadowOffset": Object {
        "height": 1,
        "width": 1,
      },
      "shadowOpacity": 0.2,
      "shadowRadius": 0.5,
    }
  }
  testID="cardHero"
>
  <View
    style={
      Object {
        "backgroundColor": "rgba(0, 0, 0, 0.35)",
        "borderRadius": 20,
        "padding": 8,
        "position": "absolute",
        "right": 10,
        "top": 10,
        "zIndex": 9,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 12,
          },
          Object {
            "color": "#fff",
            "fontSize": 18,
          },
          Object {
            "fontFamily": "anticon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <Image
    source={
      Object {
        "uri": "test",
      }
    }
    style={
      Object {
        "borderTopLeftRadius": 10,
        "borderTopRightRadius": 10,
        "height": 180,
        "width": 355,
      }
    }
  />
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FAA367",
        "flexDirection": "row",
        "justifyContent": "center",
        "opacity": 1,
        "padding": 10,
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#333",
            "fontSize": 18,
          },
          undefined,
          Object {
            "fontFamily": "anticon",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "color": "#333",
          "marginLeft": 8,
        }
      }
    >
      Favoritar
    </Text>
  </View>
  <Text
    style={
      Object {
        "alignSelf": "center",
        "fontSize": 20,
        "fontWeight": "bold",
        "marginVertical": 20,
        "maxWidth": 140,
        "textAlign": "center",
      }
    }
  >
    Wolverine
  </Text>
</View>
`;
